src/rubrix/client/datasets.py

```diff
@@ -802,44 +808,84 @@ def prepare_for_training(self) -> "datasets.Dataset":
 
 
         """
-        import datasets
+        if framework is "transformers":
 
-        has_annotations = False
-        for rec in self._records:
-            if rec.annotation is not None:
-                has_annotations = True
-                break
-
-        if not has_annotations:
-            return datasets.Dataset.from_dict({})
-
-        class_tags = ["O"]
-        class_tags.extend(
-            [
-                f"{pre}-{label}"
-                for label in sorted(self.__all_labels__())
-                for pre in ["B", "I"]
-            ]
-        )
-        class_tags = datasets.ClassLabel(names=class_tags)
+            import datasets
 
-        def spans2iob(example):
-            r = TokenClassificationRecord(
-                text=example["text"],
-                tokens=example["tokens"],
-                annotation=self.__entities_to_tuple__(example["annotation"]),
+            has_annotations = False
+            for rec in self._records:
+                if rec.annotation is not None:
+                    has_annotations = True
+                    break
+
+            if not has_annotations:
+                return datasets.Dataset.from_dict({})
+
+            class_tags = ["O"]
+            class_tags.extend(
+                [
+                    f"{pre}-{label}"
+                    for label in sorted(self.__all_labels__())
+                    for pre in ["B", "I"]
+                ]
             )
-            return class_tags.str2int(r.spans2iob(r.annotation))
+            class_tags = datasets.ClassLabel(names=class_tags)
 
-        ds = (
-            self.to_datasets()
-            .filter(self.__only_annotations__)
-            .map(lambda example: {"ner_tags": spans2iob(example)})
-        )
-        new_features = ds.features.copy()
-        new_features["ner_tags"] = [class_tags]
+            def spans2iob(example):
+                r = TokenClassificationRecord(
+                    text=example["text"],
+                    tokens=example["tokens"],
+                    annotation=self.__entities_to_tuple__(example["annotation"]),
+                )
+                return class_tags.str2int(r.spans2iob(r.annotation))
+
+            ds = (
+                self.to_datasets()
+                .filter(self.__only_annotations__)
+                .map(lambda example: {"ner_tags": spans2iob(example)})
+            )
+            new_features = ds.features.copy()
+            new_features["ner_tags"] = [class_tags]
+
+            return ds.cast(new_features)
 
-        return ds.cast(new_features)
+        elif framework == "spacy":
+
+            import spacy
+            from spacy.tokens import DocBin
+
+            nlp = spacy.blank("es")
```

Comment:
I think ideally the language should be an argument to the `prepare_for_training` method to get the right `Vocab` (which affects the lexemes). But since I don't know if this could be important for the training or not, I would leave it for a second iteration and choose as default the `en` language. Simply because it is the most common one.

Reply:
Would it be difficult to add the language as an argument to the `prepare_for_training` method? I think this is an important aspect of this feature.

Reply:
Do make sense for a more general argument to select the framework model? for example `en_core_news_sm` for spaCy or a hf model URI for hf transformers.

This could help by loading the model tokenizer and providing as result the tokenized text according to the model.

Reply:
Regarding the HF tokenizer, not sure if I would go down this rabbit hole. There are a lot of options you can pass on to the tokenizer and its call method (truncation, padding, what to return, ...). I feel like this would need a bit more thought to cover all use cases and to have a more or less clean signature.

 Regarding the spacy language: i was reading a bit and different languages use different tokenizers
 ```python
 nlp_en, nlp_es = spacy.blank("en"), spacy.blank("es")
 list(nlp_en("don't"))  # [do, n't]
 list(nlp_es("don't"))  # [don't]
 ```
 So we definitely should include a `lang` argument in the method.

Reply:
I agree with @dcfidalgo on this wrt hugging face tokenizers, these should be an external step done by the users.

For spaCy it seems we do need at least the language (if we are to use blank) or the model_name. For the model name I am not sure, we need to understand what minimal components are needed/used for the docbin creation, if there are no trainable components we should be good with blank and the language code.

Reply:
Regarding future features, could the language be a potential candidate to be included as part of record/dataset info?

Anyway, I assume we can leverage most of the use cases by including the lang as an argument. 

It will take effect for other than spaCy frameworks?

Reply:
Regarding the options you can tune for the HF tokenizer, I think the same can be applied to the spaCy pipeline. You would like to customize your spacy tokenizer before generating the training data.

Reply:
Yes, you are right. 

I think that for spaCy we should pass the pipeline as arg instead of the `str` with the lang name or model name:

```python
def blabla( nlp: Language = spacy.blank("en")):

```

For Hugging Face they do not have a specialized format for training data so it's a bit different, and I don't have a clear view of how to deal with the relationship between the tokenizer and the ents defined in Rubrix

Reply:
The other thing is if we need to use the Language object at all. We already have the tokenized text, the clear text, and the spans, or does this conflict with the recent issue about whitespaces? Maybe better discuss this over a call.

Reply:
So, if the lang is None, we use `nlp = spacy.blank("en")`? Also, are we forcing the user to use exactly the language code Spacy supports? en for English, es for Spanish, etc.

Having two methods also seems a little bit verbose for me, even though I am not a particular fan of passing a string parameter. I can proceed on this matter in both ways. 

Sorry for the late response, by the way. I hope you are all doing fine in that heat wave, especially the Madrilians. I'm going home this Friday, so I'll have my share as well 🥶

Reply:
The idea here is more or less following workflow:
```python
import spacy
nlp = spacy.blank("en")

my_text = "This is just a 'don't do it' example"
tokens = list(map(str, nlp(my_text)))
rec = rb.TokenClassificationRecord(text=my_text, tokens=tokens, annotation=[("do", 16, 18)])
ds = rb.DatasetForTokenClassification([rec])

docbin = ds.prepare_for_training(framework="spacy", lang=nlp)
```
Maybe, for now, I would omit a default value for `lang` to invite the user to think a bit about this argument, what do you think @frascuchon @dvsrepo

Reply:
> Maybe, for now, I would omit a default value for lang to invite the user to think a bit about this argument, what do you think @frascuchon @dvsrepo

I'm not really conscious about the complete flow, but it has the sense to me since setting a default language could raise some training troubles for text language other than English.